code/game/turfs/simulated/floor_icon.dm

```diff
@@ -21,107 +24,88 @@ var/list/flooring_cache = list()
 				flooring_override = icon_state
 			else
 				icon_state = flooring.icon_base
-				if(flooring.has_base_range)
-					icon_state = "[icon_state][rand(0,flooring.has_base_range)]"
-					flooring_override = icon_state
 
 
 		// Apply edges, corners, and inner corners.
 		overlays.Cut()
 		var/has_border = 0
-		if(flooring.flags & SMOOTH_ONLY_WITH_ITSELF) // for carpets and stuff like that
-			if(isnull(set_update_icon) && (flooring.flags & TURF_HAS_EDGES))
+		if (!flooring.smooth_nothing)
+		//Check if we're actually going to do anything first
+			if (isnull(set_update_icon))
+
+				//Check the cardinal turfs
 				for(var/step_dir in cardinal)
 					var/turf/simulated/floor/T = get_step(src, step_dir)
-					if(!istype(T) || !T.flooring || T.flooring.name != flooring.name)
 
+					var/is_linked = test_link(T)
+
+
+
+
+					//Alright we've figured out whether or not we smooth with this turf
+					if (!is_linked)
 						has_border |= step_dir
-						if ((flooring.flags & TURF_EDGES_EXTERNAL))
 
+						//Now, if we don't, then lets add a border
+						if ((flooring.flags & TURF_EDGES_EXTERNAL))
 							var/odir = turn(step_dir, 180)
 							var/image/I = get_flooring_overlay("[flooring.icon_base]-ext-edge-[odir]", "[flooring.icon_base]_edges", odir, TRUE)
 							overlays |= I
 						else
 							overlays |= get_flooring_overlay("[flooring.icon_base]-edge-[step_dir]", "[flooring.icon_base]_edges", step_dir)
 
+				//By doing &15 we only take the first four bits, which represent NORTH, SOUTH, EAST, WEST
+				has_smooth = ~(has_border & 15)
 
 
-				if ((flooring.flags & TURF_USE0ICON) && has_border)
-					icon_state = flooring.icon_base+"0"
+			//We can only have inner corners if we're smoothed with something
+			if (has_smooth)
+				if(flooring.flags & TURF_HAS_INNER_CORNERS)
+
+					//Quick way to check if we're smoothed with both north and east
+					if((has_smooth & NORTHEAST) == NORTHEAST)
```

Comment:
`== NORTHEAST` is not needed. See above `if(flooring.flags & TURF_HAS_INNER_CORNERS)` for example on how to use binary ops to work with flags. This applies to rest of these conditions too.

Reply:
it is needed, otherwise it returns true for either north or east. 
This check ensures its only true for both